src/rust/ide/src/model/execution_context.rs

```diff
@@ -249,129 +236,65 @@ pub struct AttachedVisualization {
-/// for the context.
-///
-/// It implements internal mutability pattern, so it may be shared between different
-/// controllers.
-#[derive(Debug)]
-pub struct ExecutionContext {
-    logger:Logger,
-    /// A name of definition which is a root call of this context.
-    pub entry_point:MethodPointer,
-    /// Local call stack.
-    stack:RefCell<Vec<LocalCall>>,
-    /// Set of active visualizations.
-    visualizations: RefCell<HashMap<VisualizationId,AttachedVisualization>>,
-    /// Storage for information about computed values (like their types).
-    pub computed_value_info_registry: ComputedValueInfoRegistry,
-}
+/// Execution Context Model API.
+pub trait API : Debug {
+    /// Obtain the method pointer to the method of the call stack's top frame.
+    fn current_method(&self) -> MethodPointer;
 
-impl ExecutionContext {
-    /// Create new execution context
-    pub fn new(logger:impl Into<Logger>, entry_point:MethodPointer) -> Self {
-        let logger                       = logger.into();
-        let stack                        = default();
-        let visualizations               = default();
-        let computed_value_info_registry = default();
-        Self {logger,entry_point,stack,visualizations, computed_value_info_registry }
-    }
+    /// Get the information about the given visualization. Fails, if there's no such visualization
+    /// active.
+    fn visualization_info(&self, id:VisualizationId) -> FallibleResult<Visualization>;
 
-    /// Push a new stack item to execution context.
-    pub fn push(&self, stack_item:LocalCall) {
-        self.stack.borrow_mut().push(stack_item);
-        self.computed_value_info_registry.clear();
-    }
+    /// Get the information about all the active visualizations.
+    fn all_visualizations_info(&self) -> Vec<Visualization>;
 
-    /// Pop the last stack item from this context. It returns error when only root call
-    /// remains.
-    pub fn pop(&self) -> FallibleResult<LocalCall> {
-        let ret = self.stack.borrow_mut().pop().ok_or_else(PopOnEmptyStack)?;
-        self.computed_value_info_registry.clear();
-        Ok(ret)
-    }
+    /// Returns IDs of all active visualizations.
+    fn active_visualizations(&self) -> Vec<VisualizationId>;
 
-    /// Obtain the method pointer to the method of the call stack's top frame.
-    pub fn current_method(&self) -> MethodPointer {
-        if let Some(top_frame) = self.stack.borrow().last() {
-            top_frame.definition.clone()
-        } else {
-            self.entry_point.clone()
-        }
-    }
+    /// Get the registry of computed values.
+    fn computed_value_info_registry(&self) -> &ComputedValueInfoRegistry;
+
+    /// Get all items on stack.
+    fn stack_items<'a>(&'a self) -> Box<dyn Iterator<Item=LocalCall> + 'a>;
+
+    /// Push a new stack item to execution context.
+    fn push(&self, stack_item:LocalCall) -> BoxFuture<FallibleResult<()>>;
```

Comment:
recheck if any of methods here could sensibly use static boxed future

Reply:
To reduce possible future breakages we decided to leave the API as-is.
Could be revisited if use case appears.